zookeeper/source/zk_nn_client.cc

```diff
@@ -0,0 +1,939 @@
+		std::uint64_t blocksize = request.blocksize();
+		std::uint32_t replication = request.replication();
+		std::uint32_t createflag = request.createflag();
+
+		if (file_exists(path)) {
+			// TODO solve this issue of overwriting files
+			LOG(ERROR) << CLASS_NAME << "File already exists";
+			return 0;
+		}
+
+		// If we need to create directories, do so
+		if (create_parent) {
+			std::string directory_paths = "";
+			auto split_path = split(path, '/');
+			LOG(INFO) << CLASS_NAME << split_path.size();
+			for (int i = 0; i < split_path.size() - 1; i++) {
+				directory_paths += split_path[i];
+			}
+			// try and make all the parents
+			if (!mkdir_helper(directory_paths, true))
+				return 0;
+		}
+
+		// Now create the actual file which will hold blocks
+		FileZNode znode_data;
+		znode_data.length = 0;
+		znode_data.under_construction = UNDER_CONSTRUCTION;
+		// http://stackoverflow.com/questions/19555121/how-to-get-current-timestamp-in-milliseconds-since-1970-just-the-way-java-gets
+		struct timeval tp;
+		gettimeofday(&tp, NULL);
+		uint64_t mslong = (uint64_t) tp.tv_sec * 1000L + tp.tv_usec / 1000; //get current timestamp in milliseconds
+		znode_data.access_time = mslong;
+		znode_data.modification_time = mslong;
+		strcpy(znode_data.owner, owner.c_str());
+		strcpy(znode_data.group, owner.c_str());
+		znode_data.replication = replication;
+		znode_data.blocksize = blocksize;
+		znode_data.filetype = 2;
+
+		// if we failed, then do not set any status
+		if (!create_file_znode(path, &znode_data))
+			return 0;
+
+		HdfsFileStatusProto* status = response.mutable_fs();
+		set_file_info(status, path, znode_data);
+
+		return 1;
+	}
+
+	void ZkNnClient::complete(CompleteRequestProto& req, CompleteResponseProto& res) {
+
+		// TODO: Completion makes a few guarantees that we should handle
+
+		int error_code;
+		// change the under construction bit
+		const std::string& src = req.src();
+		FileZNode znode_data;
+		read_file_znode(znode_data, src);
+		znode_data.under_construction = FILE_COMPLETE;
+		std::vector<std::uint8_t> data(sizeof(znode_data));
+		file_znode_struct_to_vec(&znode_data, data);
+		if (!zk->set(ZookeeperPath(src), data, error_code)) {
+			// TODO : handle erro
+		}
+		res.set_result(true);
+	}
+
+	/**
+	 * Rename a file in the zookeeper filesystem
+	 */
+	void ZkNnClient::rename(RenameRequestProto& req, RenameResponseProto& res) {
+		if(!rename_file(req.src(), req.dst())) {
+			res.set_result(false);
+		}
+		res.set_result(true);
+	}
+
+	// ------- make a directory
+
+	/**
+	 * Set the default information for a directory znode
+	 */
+	void ZkNnClient::set_mkdir_znode(FileZNode* znode_data) {
+		znode_data->length = 0;
+		struct timeval tp;
+		gettimeofday(&tp, NULL);
+		// http://stackoverflow.com/questions/19555121/how-to-get-current-timestamp-in-milliseconds-since-1970-just-the-way-java-gets
+		uint64_t mslong = (uint64_t) tp.tv_sec * 1000L + tp.tv_usec / 1000; //get current timestamp in milliseconds
```

Comment:
I can do this, I will need to use this to finish the wait_for_acks implementation